src/Microsoft.AspNetCore.Routing/RouteValuesBasedEndpointFinder.cs

```diff
@@ -56,14 +62,26 @@ public IEnumerable<Endpoint> FindEndpoints(RouteValuesBasedEndpointFinderContext
                 .Select(match => (MatcherEndpoint)match.Entry.Data);
         }
 
+        private void HandleChange()
+        {
+            // rebuild the matches
+            BuildOutboundMatches();
+
+            // re-register the callback as the change token is one time use only and a new change token
+            // is produced every time
+            Extensions.Primitives.ChangeToken.OnChange(
+                        () => _endpointDataSource.ChangeToken,
+                        () => HandleChange());
+        }
+
         private void BuildOutboundMatches()
         {
-            var (allOutboundMatches, namedOutboundMatches) = GetOutboundMatches();
-            _namedMatches = GetNamedMatches(namedOutboundMatches);
-            _allMatchesLinkGenerationTree = new LinkGenerationDecisionTree(allOutboundMatches.ToArray());
+            var (allMatches, namedMatches) = GetOutboundMatches();
+            _namedMatches = GetNamedMatches(namedMatches);
+            _allMatchesLinkGenerationTree = new LinkGenerationDecisionTree(allMatches.ToArray());
```

Comment:
Add a comment here that this is OK because no consumer will read both of these fields. Otherwise you need a lock.